docs/v3/codecs/sharding-indexed/v1.0.rst

```diff
@@ -0,0 +1,224 @@
+
+Copyright 2022-Present `Zarr core development team
+<https://github.com/orgs/zarr-developers/teams/core-devs>`_. This work
+is licensed under a `Creative Commons Attribution 3.0 Unported License
+<https://creativecommons.org/licenses/by/3.0/>`_.
+
+----
+
+
+Abstract
+========
+
+This specification defines an implementation of the Zarr codec specification 
+for sharding.
+
+Sharding logically splits chunks ("shards") into sub-chunks ("inner chunks") 
+that can be individually compressed and accessed. This allows to colocate 
+multiple chunks within one storage object, bundling them in shards.
+
+
+Motivation
+==========
+
+In many cases, it becomes inefficient or impractical to store a large number of
+chunks in single files or objects due to the design constraints of the
+underlying storage. For example, the file block size and maximum inode number
+restrict the usage of numerous small files for typical file systems, also cloud
+storage such as S3, GCS, and various distributed filesystems do not efficiently
+handle large numbers of small files or objects.
+
+Increasing the chunk size works only up to a certain point, as chunk sizes need
+to be small for read and write efficiency requirements, for example to stream
+data in browser-based visualization software.
+
+Therefore, chunks may need to be smaller than the minimum size of one storage
+key. In those cases, it is efficient to store objects at a more coarse
+granularity than reading chunks.
+
+**Sharding solves this by allowing to store multiple chunks in one storage key,
+which is called a shard**:
+
+.. image:: sharding.png
+
+
+Document conventions
+====================
+
+Conformance requirements are expressed with a combination of descriptive
+assertions and [RFC2119]_ terminology. The key words "MUST", "MUST NOT",
+"REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY",
+and "OPTIONAL" in the normative parts of this document are to be interpreted as
+described in [RFC2119]_. However, for readability, these words do not appear in
+all uppercase letters in this specification.
+
+All of the text of this specification is normative except sections explicitly
+marked as non-normative, examples, and notes. Examples in this specification are
+introduced with the words "for example".
+
+
+Configuration parameters
+========================
+
+Sharding can be configured per array in the :ref:`array-metadata` as follows:
+
+.. code-block::
+    {
+      codecs: [
+        {
+          "name": "sharding_indexed"
+          "configuration": {
+            "chunk_shape": [
+                32,
+                32
+            ],
+            "codecs": [
+              {
+                "name": "gzip",
+                "configuration": {
+                  "level": 1
+                }
+              }
+            ]
+          }
+        }
+      ]
+    }
+
+``chunk_shape``
```

Comment:
Rather than `chunk_shape` here, it would be nice to use `chunk_grid` exactly as in the top-level array metadata.  That way this can accommodate a rectilinear chunk grid for the inner chunks as well.  However, I realized there is a complication:

Ideally we could support a fully general rectilinear sub-chunk grid, e.g.:


```json
{
  "chunk_grid": {
    "type": "rectilinear",
    "configuration": {
      "chunk_shape": [[15, 35]]
    }
  },
  "codecs": [{
    "name": "sharding_indexed",
    "configuration": {
      "chunk_grid": {
        "type": "rectilinear",
        "chunk_shape": [[5, 10, 15, 20]]
      },
      "codecs": [{"name": "gzip"}]
    }
  }]
}
```

Here, the `chunk_grid` specified for `sharding_indexed` applies to the overall array, rather than the individual chunk.  The chunk boundaries must be aligned with the top-level chunk boundaries.  This allows us to have a different sub-chunk grid for each shard, which provides the most flexibility.

However, the metadata structure would make it more natural for the `chunk_grid` specified for `sharding_indexed` to apply only to the shard itself, which would mean the same chunk grid has to be used for every shard, and means that rectilinear chunks don't really interoperate with sharding well.

There is also another problem with having the `chunk_grid` specified for `sharding_indexed` apply to the overall array: it requires that the codec know which region of the overall array it is operating on.  But if the `transpose` codec, for example, is applied before `sharding_indexed`, then it is not clear how to provide `sharding_indexed` with information about its position in the overall array.

Reply:
I am happy to change the metadata to the `chunk_grid` object. Since there isn't a specific proposal for rectilinear arrays yet, I think we don't need to decide this yet. For now, I would add that the inner `chunk_grid` needs to be `regular`.

> However, the metadata structure would make it more natural for the `chunk_grid` specified for `sharding_indexed` to apply only to the shard itself, which would mean the same chunk grid has to be used for every shard, and means that rectilinear chunks don't really interoperate with sharding well.

It almost seems like the `chunk_grid` configuration for rectilinear grids would need to be added to the shard index.

Reply:
I think you have a good point --- in practice if using rectilinear chunk grids, it would be most useful to be able to specify the chunking of each shard as part of the shard, rather than in the top-level metadata.  That would require a different shard format, though.

Using the same non-regular rectilinear chunking for every shard is unlikely to be useful, but specifying separate rectilinear chunking for each shard in the top-level metadata would be rather awkward and doesn't fit well with how codecs are supposed to work.

Therefore, I withdraw this suggestion; I think it is fine with just a `chunk_shape` member as you have it, rather than using `chunk_grid`.

Reply:
I actually think keeping the `chunk_grid` in is a good idea for later extensibility and consistent with the v3 mission. If we would adding support for rectilinear grids through the binary format, it would need to be specified somewhere. Also, we might come up with other grids that make more sense for sharding.

Reply:
At least the current rectilinear grid proposal puts all of the partition points in the array metadata itself, e.g. `{"name": "rectilinear", "chunk_shape": [[100, 100, 50, 100], [150, 150, 25, 100], [200, 300, 100, 200]]}`.  Whereas for sharding we would perhaps just specify `{"name": "rectilinear"}` but then rely on a quite different binary format to actually specify the grid.    Additionally, the type of grid could also be specified in the binary format, as you could imagine that even with a regular grid, you might want different shards to use a different chunk shape.  Given these differences I'm not sure it makes sense to try to tie the shard chunking to the top-level array chunking.  Instead there could just be a different codec name.

Reply:
Fine with me.